src/Features/CSharp/Portable/UseRecursivePatterns/AnalyzedNodeFactory.cs

```diff
@@ -0,0 +1,421 @@
+                {
+                    return new IndexEvaluation(VisitInput(op.Instance), indexer, (int)constant.ConstantValue.Value, op.Syntax);
+                }
+
+                return null;
+            }
+
+            static Evaluation? VisitNullableTypeMemberAccess(IPropertyReferenceOperation op)
+            {
+                if (op.Property.ContainingType.OriginalDefinition.SpecialType != SpecialType.System_Nullable_T)
+                    return null;
+
+                // Special cases for a `Nullable<T>` receiver:
+                return op.Property.Name switch
+                {
+                    // (1) Checking `HasValue` is interpreted as a null check
+                    nameof(Nullable<int>.HasValue) => new NotNull(VisitInput(op.Instance), op.Syntax),
+                    // (2) Skipping `Value` access since in patterns this is implicit
+                    nameof(Nullable<int>.Value) => VisitInput(op.Instance),
+                    _ => null
+                };
+            }
+        }
+
+        private static INamedTypeSymbol? GetITupleType(IOperation op)
+        {
+            return op.SemanticModel.Compilation.GetTypeByMetadataName("System.Runtime.CompilerServices.ITuple");
+        }
+
+        private static AnalyzedNode? VisitNotOperator(IUnaryOperation op)
+        {
+            var operand = Visit(op.Operand);
+            if (operand is null || operand is OperationEvaluation)
+                return null;
+
+            return Not.Create(operand);
+        }
+        private static AnalyzedNode? VisitBinary(IOperation left, IOperation right, bool disjunctive, Evaluation? input)
+        {
+            var leftNode = VisitCore(left, input);
+            if (leftNode is null)
+                return null;
+
+            var rightNode = VisitCore(right, input);
+            if (rightNode is null)
+                return null;
+
+            return Sequence.Create(disjunctive, leftNode, rightNode);
+        }
+
+        private static AnalyzedNode? VisitConditionalOrOperator(IBinaryOperation op)
+            => VisitBinary(op.LeftOperand, op.RightOperand, disjunctive: true, input: null);
+
+        private enum ConstantResult { None, Left, Right }
+
+        private static (ConstantResult, IOperation) DetermineConstant(IBinaryOperation operation)
+        {
+            return (operation.LeftOperand, operation.RightOperand) switch
+            {
+                var (_, v) when IsConstant(v, out var op) => (ConstantResult.Right, op),
+                var (v, _) when IsConstant(v, out var op) => (ConstantResult.Left, op),
+                _ => default,
+            };
+        }
+
+        private static bool IsConstant(IOperation operation, out IOperation constant)
+        {
+            constant = Unwrap(operation);
+            return constant.ConstantValue.HasValue;
+
+            static IOperation Unwrap(IOperation operation)
+            {
+                while (operation is IConversionOperation op)
```

Comment:
should we only be considering non-user-defined conversions?  a user defined conversion would not preserve constant-ness.

also, as we wrap with IConversions, don't we still keep the constant value flowing through the iop tree?  this seems a little suspect.